db/db_log_iter_test.cc

```diff
@@ -168,17 +170,22 @@ TEST_F(DBTestXactLogIterator, TransactionLogIteratorCheckWhenArchive) {
     ASSERT_OK(dbfull()->Put(WriteOptions(), "key4", DummyString(1024)));
     ASSERT_OK(dbfull()->Flush(FlushOptions()));
 
+    callback_hit.StoreRelaxed(false);
     ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
         "WalManager::PurgeObsoleteFiles:1", [&](void*) {
           auto iter = OpenTransactionLogIter(0);
           ExpectRecords(4, iter);
+          callback_hit.StoreRelaxed(true);
         });
 
     ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
     ASSERT_OK(dbfull()->Flush(FlushOptions(), cf));
-
     delete cf;
+    // Normally hit several times; WART: perhaps more in parallel after flush
```

Comment:
Oh I think I know what's going on. The callback is actually not supposed to be called multiple times in each iteration, only after the final (third) flush; the reason it does get called repeatedly is that we don't clear the callback at the end of the loop body (so e.g. during the second iteration, the first callback is still there and gets triggered after all three flushes). This is an accident waiting to happen because the callback has a db pointer, which can dangle during the next iteration or after exiting the loop.

Reply:
`this` is captured, not a DB pointer. I get the same number of callback hits regardless of whether disabling callback processing at the end of each iteration. If/when some background work is deferred to closing time, we wouldn't want to disable it prematurely, assuming that's a valid scenario to test. (In my estimation it is.)

Reply:
> `this` is captured, not a DB pointer.

Ah, you're right. I misread the code.